### ISSUE TITLE
Enabling profiling causes random UnableToResolvePhysicalConnection exceptions

### ISSUE THREAD

**Issue by TheCloudlessSky** [StackExchange/StackExchange.Redis#280]
I'm using StackExchange.Redis (1.0.481) successfully for a long time. I'm looking into adding profiling to report in our production environment. I have an MVC 5.2.3 application that's wired up via Ninject (DI). I've created an HttpModule that begins/ends profiling during the `BeginRequest` and `EndRequest` events (pretty much copied straight from the Docs):

``` csharp
public class HttpContextRedisProfiler : IProfiler
{
    private const string key = "HttpContextRedisProfiler.Context";

    public object GetContext()
    {
        var httpContext = HttpContext.Current;
        if (httpContext == null) return null;

        return httpContext.Items[key];
    }

    public object CreateContextForCurrentRequest()
    {
        var httpContext = HttpContext.Current;
        if (httpContext == null) return null;

        var result = new object();
        httpContext.Items[key] = result;
        return result;
    }
}

public class MeasureRedisModule : IHttpModule
{
    private readonly HttpContextRedisProfiler profiler;
    private readonly ConnectionMultiplexer connectionMultiplexer;

    public MeasureRedisModule(HttpContextRedisProfiler profiler, ConnectionMultiplexer connectionMultiplexer)
    {
        this.profiler = profiler;
        this.connectionMultiplexer = connectionMultiplexer;
    }

    public void Init(HttpApplication context)
    {
        context.BeginRequest += OnBeginRequest;
        context.EndRequest += OnEndRequest;
    }

    private void OnBeginRequest(object sender, EventArgs e)
    {
        var context = profiler.CreateContextForCurrentRequest();
        if (context != null)
        {
            connectionMultiplexer.BeginProfiling(context);
        }
    }

    private void OnEndRequest(object sender, EventArgs e)
    {
        var context = profiler.GetContext();
        if (context == null) return;

        var profiledCommands = connectionMultiplexer.FinishProfiling(context);

        foreach (var result in profiledCommands)
        {
            // Do something with the commands (e.g. log them to our metrics API).
        }
    }

    public void Dispose()
    {
    }
}

// Setup

var config = new ConfigurationOptions();
config.EndPoints.Add("localhost:6379");
// Increase the timeout so that AOF writes do not cause exceptions:
// https://github.com/StackExchange/StackExchange.Redis/issues/83
config.SyncTimeout = 2000;
var connectionMultiplexer = ConnectionMultiplexer.Connect(config);
var profiler = new HttpContextRedisProfiler();
connectionMultiplexer.RegisterProfiler(profiler);
```

However, when I run the application locally, I'm getting random `UnableToResolvePhysicalConnection` exceptions (not every request causes it). These exceptions do not occur with profiling turned off (or not calling `BeginProfiling`/`FinishProfiling`).

I've read [Marc's comments](https://github.com/StackExchange/StackExchange.Redis/issues/68#issuecomment-48570718) about why this exception occurs. However, it doesn't make sense that enabling profiling causes this.

Specifically, the stack trace for the exception that seems to occur most often is:

```
[RedisConnectionException: UnableToResolvePhysicalConnection on EVAL]
   StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl(Message message, ResultProcessor`1 processor, ServerEndPoint server) in c:\TeamCity\buildAgent\work\3ae0647004edff78\StackExchange.Redis\StackExchange\Redis\ConnectionMultiplexer.cs:1935
   StackExchange.Redis.RedisBase.ExecuteSync(Message message, ResultProcessor`1 processor, ServerEndPoint server) in c:\TeamCity\buildAgent\work\3ae0647004edff78\StackExchange.Redis\StackExchange\Redis\RedisBase.cs:80
   StackExchange.Redis.RedisDatabase.ScriptEvaluate(String script, RedisKey[] keys, RedisValue[] values, CommandFlags flags) in c:\TeamCity\buildAgent\work\3ae0647004edff78\StackExchange.Redis\StackExchange\Redis\RedisDatabase.cs:866
   StackExchange.Redis.RedisDatabase.ScriptEvaluate(LuaScript script, Object parameters, CommandFlags flags) in c:\TeamCity\buildAgent\work\3ae0647004edff78\StackExchange.Redis\StackExchange\Redis\RedisDatabase.cs:884
   MyApp.Web.Infrastructure.RedisRateLimitSetUserLastActivityDecorator.Set(User user, DateTime now) in d:\Repositories\MyApp\src\MyApp.Web\Infrastructure\RedisRateLimitSetUserLastActivityDecorator.cs:53
... Snipped the stack trace for our app ...
```

The `RedisRateLimitSetUserLastActivityDecorator` simply evaluates a script synchronously:

``` csharp
// Use a script for atomicity.
var script = LuaScript.Prepare(@"
local windowExists = redis.call('exists', @key)
if windowExists == 1 then
    return false
else
    redis.call('set', @key, 1)
    redis.call('expire', @key, @expireSeconds)
    return true
end
");

var key = "UsersLastActivityRateLimit/v1/" + user.Id;

var rawResult = redis.ScriptEvaluate(script, new { key = key, expireSeconds = 60 });
var rateLimitWindowExists = (bool)rawResult;
// ...snip...
```

Any idea why these exceptions are occurring locally only when profiling is enabled? I can test with a non-Windows instance of Redis if required.

**Comment by RegiSV2** [comment]
I'ce been experiencing similar issues when tried to enable profiling on one of our .NET Core production servers. I'm not sure though if it is the same issue TheCloudlessSky is writing about.

I've added event handler to ConnectionMultiplexer.InternalError event and logged all the errors, then examined errors that occured at the same time when UnableToResolvePhysicalConnection exceptions where encountered in my code. It occured that all exceptions from InternalError where nearly the same:

`WriteQueue. Interactive. System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at StackExchange.Redis.ProfileStorage.SetRequestSent()
   at StackExchange.Redis.PhysicalBridge.WriteMessageDirect(PhysicalConnection tmp, Message next)
   at StackExchange.Redis.PhysicalBridge.WriteQueue(Int32 maxWork) `

It seems like this exception is occuring here:
https://github.com/StackExchange/StackExchange.Redis/blob/32807af6d64220f93c182fdc79d5969087f1d9d9/StackExchange.Redis/StackExchange/Redis/ProfileStorage.cs#L83
And then it is caught in WriteQueue loop, causing to finish async Redis operation whith errors and throw an exception in caller code:
https://github.com/StackExchange/StackExchange.Redis/blob/32807af6d64220f93c182fdc79d5969087f1d9d9/StackExchange.Redis/StackExchange/Redis/PhysicalBridge.cs#L619

So, ProfileStorage.SetEnqueued is sometimes called multiple times, although it is expected to be called only once, which seems like a bug to me. The easy fix would be to ignore subsequent calls in SetEnqueued instead of throwing InvalidOperationExceptions, but maybe the bug is somewhere else.

**Comment by gliljas** [comment]
Yup, it seems to happen once per year. Encountered exactly what @RegiSV2 explained. I have disabled profiling now. The UnableToResolvePhysicalConnection (mostly for PUBLISH commands) exceptions weren't thrown (at least visibly) before.

**Comment by NickCraver** [comment]
Fixed in #769, thanks @gliljas!
